core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/snapshot/DefaultProgressSnapshot.java

```diff
@@ -0,0 +1,174 @@
+        }
+        Validate.paramNotNull(builder.transferredBytes, "byteTransferred");
+        this.transferredBytes = Validate.isNotNegative(builder.transferredBytes, "transferredBytes");
+        this.totalBytes = builder.totalBytes;
+
+        if (builder.startTime.isPresent()) {
+            Instant currentTime = Instant.now();
+            Validate.isTrue(currentTime.isAfter(builder.startTime.get()),
+                            "currentTime (%s) must not be before startTime (%s)",
+                            currentTime, builder.startTime.get());
+        }
+
+        this.startTime = builder.startTime;
+    }
+
+    @Override
+    public long transferredBytes() {
+        return this.transferredBytes;
+    }
+
+    @Override
+    public Optional<Instant> startTime() {
+        return this.startTime;
+    }
+
+    @Override
+    public Optional<Duration> elapsedTime() {
+        return this.startTime.isPresent() ? Optional.of(Duration.between(startTime.get(), Instant.now())) : Optional.empty();
+    }
+
+    @Override
+    public Optional<Duration> estimatedTimeRemaining() {
+        if (!elapsedTime().isPresent() || !remainingBytes().isPresent()) {
+            return Optional.empty();
+        }
+
+        long remainingTime = remainingBytes().getAsLong() * elapsedTime().get().toMillis() / transferredBytes;
+        return Optional.of(Duration.ofMillis(remainingTime));
+
+    }
+
+    @Override
+    public OptionalDouble averageBytesPer(TimeUnit timeUnit) {
+        if (!this.elapsedTime().isPresent()) {
+            return OptionalDouble.empty();
+        }
+
+        return this.elapsedTime().get().equals(Duration.ZERO) ? OptionalDouble.of(1.0) :
+               OptionalDouble.of((double) this.transferredBytes / timeUnit.convert(elapsedTime().get().toMillis(), timeUnit));
+    }
+
+    @Override
+    public OptionalLong totalBytes() {
+        return totalBytes == null ? OptionalLong.empty() : OptionalLong.of(totalBytes);
+    }
+
+    @Override
+    public OptionalDouble ratioTransferred() {
+        if (totalBytes == null) {
+            return OptionalDouble.empty();
+        }
+        return totalBytes == 0 ? OptionalDouble.of(1.0) : OptionalDouble.of(transferredBytes / totalBytes.doubleValue());
+    }
+
+    @Override
+    public OptionalLong remainingBytes() {
+        if (totalBytes == null) {
+            return OptionalLong.empty();
+        }
+        return totalBytes == 0 ? OptionalLong.of(0) : OptionalLong.of(totalBytes - transferredBytes);
+    }
+
+    @Override
+    public DefaultProgressSnapshot.Builder toBuilder() {
+        return new Builder(this);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder implements CopyableBuilder<Builder, DefaultProgressSnapshot> {
+        private long transferredBytes;
+        private Long totalBytes;
+        private Optional<Instant> startTime = Optional.empty();
```

Comment:
Same here, optional should not be used for member variables.

Reply:
Removed Optional